File: Backend/app/services/notification_service.py
from app.models.notification import Notification


def create_notification(
    db,
    recipient_type,
    recipient_name,
    notification_type,
    title,
    message,
    recipient_id=None,
    problem_id=None,
    project_id=None,
    milestone_id=None
):
    notification = Notification(
        recipient_type=recipient_type,
        recipient_id=recipient_id,
        recipient_name=recipient_name,
        problem_id=problem_id,
        project_id=project_id,
        milestone_id=milestone_id,
        notification_type=notification_type,
        title=title,
        message=message,
        is_read=False
    )

    db.add(notification)
    db.commit()
    db.refresh(notification)

    return notification

File: Frontend/src/services/notificationService.ts
import { apiClient } from './apiClient';
import { AppNotification } from '../types';

export const notificationService = {
  async getNotifications(params?: { recipient_type?: string; recipient_name?: string }): Promise<AppNotification[]> {
    const res = await apiClient.get<AppNotification[]>('/api/problems/notifications', { params });
    return res.data;
  },

  async markAsRead(notificationId: number): Promise<{ message: string; notification: AppNotification }> {
    const res = await apiClient.put<{ message: string; notification: AppNotification }>(
      `/api/problems/notifications/${notificationId}/read`
    );
    return res.data;
  },

  async createNotification(payload: {
    recipient_type: string;
    recipient_name?: string;
    notification_type: string;
    title: string;
    message: string;
    problem_id?: number;
    project_id?: number;
  }): Promise<{ message: string; notification: AppNotification }> {
    const res = await apiClient.post('/api/problems/notifications', payload);
    return res.data;
  },
};
